panos/base.py

```diff
@@ -5235,3 +5235,47 @@ def clock(self):
         fmt = "%a %b %d %H:%M:%S %Z %Y"
         text = res.text.strip()
         return datetime.strptime(text, fmt)
+
+    def plugins(self):
+        """Returns plugin information.
+
+        Each dict in the list returned has the following keys:
+            * name
+            * version
+            * release_date
+            * release_note_url
+            * package_file
+            * size
+            * platform
+            * installed
+            * downloaded
+
+        Returns:
+            list of dicts
+        """
+        # Older versions of PAN-OS do not have this command, so if we get an
+        # exception, just return None.
+        try:
+            res = self.op("<show><plugins><packages/></plugins></show>", cmd_xml=False)
+        except err.PanDeviceError:
+            return None
+
+        ans = []
+        for o in res.findall("./result/plugins/entry"):
+            ans.append(
+                {
+                    "name": o.find("./name").text,
+                    "version": o.find("./version").text,
+                    "release_date": o.find("./release-date").text,
+                    "release_note_url": (
+                        o.find("./release-note-url").text or ""
+                    ).strip(),
+                    "package_file": o.find("./pkg-file").text,
+                    "size": o.find("./size").text,
+                    "platform": o.find("./platform").text,
+                    "installed": o.find("./installed").text,
+                    "downloaded": o.find("./downloaded").text,
```

Comment:
Is there a chance any of these `o.find()` could return None, which would cause an exception when asking for `.text`?

Reply:
The only thing that could end up being `None` is the release note URL, which is why there is special handling with that specific field, as well as a new test case using a live device's response that I added to make sure the parsing is handled correctly.